### DISCUSSION TITLE
Sprint Qualifying Codes

### DISCUSSION THREAD

**Discussion by TNAKKK** [theOehrly/Fast-F1#584] [Q&A]
Hello, I'm trying to see Miami Sprint Qualify plots between Ricciardo and Tsunoda without success with the code below as it's totally outside my ability!

Can anyone tell me the correct code, please?

import fastf1 as ff1
from fastf1 import plotting
from fastf1 import utils

from matplotlib import pyplot as plt
from matplotlib.pyplot import figure

import numpy as np
import pandas as pd

# Enable the cache by providing the name of the cache folder
ff1.Cache.enable_cache('cache')

year, grand_prix, session = 2024, 'MIAMI', 'Sprint Qualifying'

quali = ff1.get_session(year, grand_prix, session)
quali.load()  # This is new with Fastf1 v.2.2

# This is how it used to be:
# laps = quali.load_laps(with_telemetry=True)

driver_1, driver_2 = 'RIC', 'TSU'

sq1_laps, sq2_laps, sq3_laps = quali.laps.split_qualifying_sessions()

# Laps can now be accessed through the .laps object coming from the session
laps_driver_1 = sq1_laps.pick_driver(driver_1)
laps_driver_2 = sq1_laps.pick_driver(driver_2)

# Select the fastest lap
fastest_driver_1 = laps_driver_1.pick_fastest()
fastest_driver_2 = laps_driver_2.pick_fastest()

# Retrieve the telemetry and add the distance column
telemetry_driver_1 = fastest_driver_1.get_telemetry().add_distance()
telemetry_driver_2 = fastest_driver_2.get_telemetry().add_distance()

# Make sure we know the team name for coloring
team_driver_1 = fastest_driver_1['Team']
team_driver_2 = fastest_driver_2['Team']

# Extract the delta time
delta_time, ref_tel, compare_tel = utils.delta_time(fastest_driver_1, fastest_driver_2)

plot_size = [15, 15]
plot_title = f"{quali.event.year} {quali.event.EventName} - {quali.name} - {driver_1} VS {driver_2}"
plot_ratios = [1, 3, 2, 1, 1, 2, 1]
plot_filename = plot_title.replace(" ", "") + ".png"

# Make plot a bit bigger
plt.rcParams['figure.figsize'] = plot_size

# Create subplots with different sizes
fig, ax = plt.subplots(7, gridspec_kw={'height_ratios': plot_ratios})

# Set the plot title
ax[0].title.set_text(plot_title)

# Delta line
ax[0].plot(ref_tel['Distance'], delta_time)
ax[0].axhline(0)
ax[0].set(ylabel=f"Gap to {driver_2} (s)")

# Speed trace
ax[1].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Speed'], label=driver_1, color='orange')
ax[1].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Speed'], label=driver_2, color='green')
ax[1].set(ylabel='Speed')
ax[1].legend(loc="lower right")

# Throttle trace
ax[2].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Throttle'], label=driver_1, color='orange')
ax[2].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Throttle'], label=driver_2, color='green')
ax[2].set(ylabel='Throttle')

# Brake trace
ax[3].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Brake'], label=driver_1, color='orange')
ax[3].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Brake'], label=driver_2, color='green')
ax[3].set(ylabel='Brake')

# Gear trace
ax[4].plot(telemetry_driver_1['Distance'], telemetry_driver_1['nGear'], label=driver_1, color='orange')
ax[4].plot(telemetry_driver_2['Distance'], telemetry_driver_2['nGear'], label=driver_2, color='green')
ax[4].set(ylabel='Gear')

# RPM trace
ax[5].plot(telemetry_driver_1['Distance'], telemetry_driver_1['RPM'], label=driver_1, color='orange')
ax[5].plot(telemetry_driver_2['Distance'], telemetry_driver_2['RPM'], label=driver_2, color='green')
ax[5].set(ylabel='RPM')

# DRS trace
ax[6].plot(telemetry_driver_1['Distance'], telemetry_driver_1['DRS'], label=driver_1, color='orange')
ax[6].plot(telemetry_driver_2['Distance'], telemetry_driver_2['DRS'], label=driver_2, color='green')
ax[6].set(ylabel='DRS')
ax[6].set(xlabel='Lap distance (meters)')

# Hide x labels and tick labels for top plots and y ticks for right plots.
for a in ax.flat:
    a.label_outer()

# Store figure
plt.savefig(plot_filename, dpi=300)
plt.show()

**Comment by TNAKKK** [comment]
Sorry just to add what I'm trying to achive is each of the comparisons below.

SQ1 RIC vs TSU

SQ2 RIC vs TSU

Many thanks!

**Comment by theOehrly** [comment]
Please, instead of just dumping a bunch of unformatted code here, please, make it a short example that shows the problem. Describe the issue you are having. Post any error message you may be getting. 
I will not attempt to debug the above code snippet. I just don't have time for that. Please try to make it easy for others to help you.

**Reply by TNAKKK** [reply]
Thanks theOehrly

I'm trying to see the comparison of the 2 laps on a plot graph. The 2 laps are from each driver's best laps from SQ2 (although I think TSU had a track limit lap time delete on his 1:28.736)

2024 MIAMI SQ2
RIC's Lap8 1:28.044
TSU's Lap8 1:28.736

Laps were found from the link below
https://fiaresultsandstatistics.motorsportstats.com/results/2024-miami-grand-prix/session-facts/744d72a7-6ba9-40e3-8ae8-e8d90a883f2d?fact=LapTime

If you could help me on how to extract them, that's much appreciated. I believe these parts are related lines but I'm sure that I'm putting wrong codes there... 

```python
year, grand_prix, session = 2024, 'Miami', 'SQ'

sq1_laps, sq2_laps, sq3_laps = quali.laps.split_qualifying_sessions()

laps_driver_1 = sq2_laps.pick_driver(driver_1)
laps_driver_2 = sq2_laps.pick_driver(driver_2)

fastest_driver_1 = laps_driver_1.pick_fastest()
fastest_driver_2 = laps_driver_2.pick_fastest()
``` 

Here are the error messages I see when trying to output a plot graph.

Traceback (most recent call last):
  File "C:\Users\tnakk\PycharmProjects\formula1_python\venv\Main.py", line 24, in <module>
    sq1_laps, sq2_laps, sq3_laps = quali.laps.split_qualifying_sessions()
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tnakk\PycharmProjects\formula1_python\venv\Lib\site-packages\fastf1\core.py", line 2679, in split_qualifying_sessions
    raise ValueError("Session is not a qualifying session!")
ValueError: Session is not a qualifying session!

Process finished with exit code 1

**Reply by theOehrly** [reply]
Ok, look, this is still not what I want to see, because this is still not an example that I can just copy and run. There are undefined variables and missing imports. The problem with that is that I need to make assumptions and fill in the blanks. First, this takes time. And second, it might change the actual outcome because I might not be running the same code as you are.

A minimal example for your problems looks like this: 

```
import fastf1

year, grand_prix, session = 2024, 'Miami', 'SQ'
quali = fastf1.get_session(year, grand_prix, session)
quali.load()
sq1_laps, sq2_laps, sq3_laps = quali.laps.split_qualifying_sessions()
```

The **full** output for that is

```
req         WARNING 	DEFAULT CACHE ENABLED! (246.4 MB) C:\Users\phisc\AppData\Local\Temp\fastf1
core           INFO 	Loading data for Miami Grand Prix - Sprint Qualifying [v3.3.4]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	No cached data found for lap_count. Loading data...
_api           INFO 	Fetching lap count data...
logger      WARNING 	Failed to load total lap count!
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
req            INFO 	Using cached data for car_data
req            INFO 	Using cached data for position_data
req            INFO 	Using cached data for weather_data
req            INFO 	Using cached data for race_control_messages
core           INFO 	Finished loading data for 20 drivers: ['1', '16', '11', '3', '55', '81', '27', '22', '10', '2', '24', '63', '23', '77', '31', '44', '14', '20', '18', '4']
Traceback (most recent call last):
  File "C:\Dateien\Code\Formula1\Fast-F1\exp_bug.py", line 6, in <module>
    sq1_laps, sq2_laps, sq3_laps = quali.laps.split_qualifying_sessions()
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Dateien\Code\Formula1\Fast-F1\src\fastf1\core.py", line 3249, in split_qualifying_sessions
    raise ValueError("Session is not a qualifying session!")
ValueError: Session is not a qualifying session!
```

Now, the error is immediately obvious (to me). You are not using the latest version of FastF1 and need to update to v3.3.5 (``pip install --upgrade fastf1``). That version contains some bug fixes for the new sprint event format this season.

If you had posted this, I would have been able to help you yesterday already, and it would have taken me maybe three minutes to see the problem and type a response. So you see why it is important to give the full information and a good example. It makes everything quicker for both of us. Keep this in mind for the next time when you run into a problem 😃

>  The 2 laps are from each driver's best laps from SQ2 (although I think TSU had a track limit lap time delete on his 1:28.736)

If Tsunodas time was deleted due to track limits, ``.pick_fastest`` will not return it by default. ``.pick_fastest(only_by_time=True)`` might return the correct lap. Else, you might need to select the lap manually by lap number.

I hope you can get everything working now.

**Comment by TNAKKK** [comment]
Appologies. I will figure it out myself. Thanks for your reply

**Reply by theOehrly** [reply]
I am happy to help you. But you need to describe the problem. From your question, it is unclear if the code raises an error or if it works, but the output is not what you expected. 

And preferably, you give a code example that is as short as possible. Then please use the option to format it as code, like so:
```
import fastf1

session = ...
```

This way, first everything is readable. And people who want to help you can easily copy and paste the code to run it.
